src/tests/profiler/native/nongcheap/nongcheap.cpp

```diff
@@ -55,20 +55,46 @@ HRESULT STDMETHODCALLTYPE NonGcHeapProfiler::ObjectAllocated(ObjectID objectId,
 
 HRESULT NonGcHeapProfiler::Shutdown()
 {
-    if (_failures > 0)
+    // Let's make sure we got the same number of objects as we got from the callback
+    // by testing the EnumerateNonGCObjects API.
+    ICorProfilerObjectEnum* pEnum = NULL;
+    HRESULT hr = pCorProfilerInfo->EnumerateNonGCObjects(&pEnum);
```

Comment:
Can you modify the test case to enable the profiler COR_PRF_MONITOR_GC events and then try doing this enumeration inside the [GarbageCollectionFinished](https://learn.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/icorprofilercallback2-garbagecollectionfinished-method) callback? I think that is one of the more likely places a profiler might want to enumerate the non-GC heap. I'm suspicious the current implementation may deadlock but even if all works well it will be good for us to confirm that scenario doesn't regress.

Reply:
Done